Tests/NIOHTTP2Tests/ConnectionStateMachineTests.swift

```diff
@@ -195,6 +217,50 @@ class ConnectionStateMachineTests: XCTestCase {
         XCTAssertTrue(self.server.fullyQuiesced)
     }
 
+    func testSimpleRequestResponseErrorFlow() {
+        let streamOne = HTTP2StreamID(1)
+        let streamTwo = HTTP2StreamID(2)
+
+        self.exchangePreamble()
+
+        // Create the stream but leave the state in idle by not passing in the required headers
+        let _ = self.server.receiveHeaders(streamID: streamOne, headers: .init(), isEndStreamSet: false)
+        var savedServerState = self.server
+            
+        // Change the state to halfOpenRemoteLocalIdle
+        let testHeaders: HPACKHeaders = [":method": "test", ":path": "test", ":scheme": "test"]
+        let _ = self.server.receiveHeaders(streamID: streamOne, headers: testHeaders, isEndStreamSet: false)
+        savedServerState = self.server
+
+        assertBadStreamStateTransition(type: .halfOpenRemoteLocalIdle, self.server.sendData(streamID: streamOne, contentLength: 0, flowControlledBytes: 0, isEndStreamSet: false))
+        self.server = savedServerState
+        
+        assertBadStreamStateTransition(type: .halfOpenRemoteLocalIdle, self.server.receivePushPromise(originalStreamID: streamOne, childStreamID: streamTwo, headers: testHeaders))
+        self.server = savedServerState
+
+        // Move state to fullyOpen
+        let testSendHeaders: HPACKHeaders = [":status": "y"]
+        assertSucceeds(self.server.sendHeaders(streamID: streamOne, headers: testSendHeaders, isEndStreamSet: false))
+        savedServerState = self.server
+
+        let testPushPromiseHeaders: HPACKHeaders = ["test": "value"]
+        assertBadStreamStateTransition(type: .fullyOpen, self.server.receivePushPromise(originalStreamID: streamOne, childStreamID: streamTwo, headers: testPushPromiseHeaders))
+        self.server = savedServerState
+
+        // Move state to halfClosedLocalPeerActive
+        assertSucceeds(self.server.sendData(streamID: streamOne, contentLength: 0, flowControlledBytes: 0, isEndStreamSet: true))
+        savedServerState = self.server
+
+        assertBadStreamStateTransition(type: .halfClosedLocalPeerActive, self.server.sendData(streamID: streamOne, contentLength: 0, flowControlledBytes: 0, isEndStreamSet: true))
+        self.server = savedServerState
+        assertBadStreamStateTransition(type: .halfClosedLocalPeerActive, self.server.sendPushPromise(originalStreamID: streamOne, childStreamID: streamTwo, headers: testPushPromiseHeaders))
+        self.server = savedServerState
+        assertBadStreamStateTransition(type: .halfClosedLocalPeerActive, self.server.receivePushPromise(originalStreamID: streamOne, childStreamID: streamTwo, headers: testPushPromiseHeaders))
+        self.server = savedServerState
+        assertBadStreamStateTransition(type: .halfClosedLocalPeerActive, self.server.sendHeaders(streamID: streamOne, headers: testPushPromiseHeaders, isEndStreamSet: false))
+        self.server = savedServerState
+    }
```

Comment:
Mind adding a test for the `CustomStringConvertible` conformance? Just try to print it.

Reply:
Done.